driver/src/orderbook/streamed/state.rs

```diff
@@ -158,11 +177,150 @@ impl State {
+                self.pending_solution = PendingSolution::AccumulatingTrades(vec![event.clone()])
+            }
+        };
+        Ok(())
+    }
+
+    fn solution_submission(
+        &mut self,
+        event: &SolutionSubmission,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        if self.tokens.get_address_by_id(0).is_none() {
+            return Err(Error::SolutionWithoutFeeToken);
+        }
+        let trades = match &mut self.pending_solution {
+            PendingSolution::AccumulatingTrades(trades) => std::mem::replace(trades, Vec::new()),
+            // This would be weird because it means that the previous solution had no trades. Not
+            // sure if we should error.
+            PendingSolution::Submitted { .. } => Vec::new(),
+        };
+        self.pending_solution = PendingSolution::Submitted {
+            batch_id: block_batch_id,
+            submitter: event.submitter,
+            burnt_fees: event.burnt_fees,
+            trades,
+        };
+        Ok(())
+    }
+
+    fn apply_pending_solution_if_needed(&mut self, block_batch_id: BatchId) {
+        let trades = match &mut self.pending_solution {
+            PendingSolution::Submitted {
+                batch_id,
+                submitter,
+                burnt_fees,
+                trades,
+            } if *batch_id < block_batch_id => {
+                // Cannot fail because we ensure there is a fee token when a solution event is
+                // received.
+                let fee_token = self.tokens.get_address_by_id(0).unwrap();
+                let balance = self.balances.entry((*submitter, fee_token)).or_default();
+                balance.balance += *burnt_fees;
+                // This looks weird but we need to call self.apply_trade outside of the match to
+                // prevent multiple mut self borrows.
+                let trades = std::mem::replace(trades, Vec::new());
+                self.pending_solution = PendingSolution::AccumulatingTrades(Vec::new());
+                trades
+            }
+            _ => return,
+        };
+        for trade in trades {
+            // Cannot fail because we ensure that the tokens exist when a trade event is
+            // received and that the order exists and doesn't get removed while it is valid.
+            self.apply_trade(
+                trade.owner,
+                trade.order_id,
+                trade.executed_sell_amount,
+                trade.executed_buy_amount,
+                block_batch_id,
+            )
+            .unwrap();
+        }
+    }
+
+    /// Fails if any tokens or order don't exist.
+    fn apply_trade(
+        &mut self,
+        user: UserId,
+        order: OrderId,
+        executed_sell_amount: u128,
+        executed_buy_amount: u128,
+        batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let order = self
+            .orders
+            .get_mut(&(user, order))
+            .ok_or(Error::UnknownOrder)?;
```

Comment:
nit: `UnknownOrder` and `UnknownToken` should include `order_id` and `token_id`.